README
lloconv
=======

This is a document converter which uses LibreOfficeKit for all the heavy
lifting.  LibreOfficeKit was formerly known as liblibreoffice, hence the name
"lloconv".

LibreOffice 4.4 and later are fully supported.  LibreOffice 4.3 works
provided you avoid filenames which contain characters which need
escaping in a URL.

To use it, just run it with the input and output filenames - the format to
convert to is determined by the extension of the output filename.  E.g.:

$ ./lloconv essay.odt essay.html

You can also fetch a document from a URL to convert:

$ ./lloconv -u https://example.org/sample.doc sample.html

Use this feature with caution - it's probably safe on https URLs you control
the content of, but fetching arbitrary documents is equivalent to automatically
opening email attachments.

The code checks for LibreOffice installed in /usr/lib/libreoffice/program
or /usr/lib64/libreoffice/program (which is where the Debian and Fedora
packages install to) and if it's not there, looks for /opt/libreoffice*/program
(which is where the .debs from libreoffice.org install to), using the highest
version if multiple are found.  You can override this by setting environmental
variable LO_PATH, either to handle an install in another location, or to force
a particular install to be used if you have more than one copy installed, e.g.:

$ LO_PATH=/opt/libreoffice5.0/program
$ export LO_PATH

Server
------

The time to initialise LibreOfficeKit can be significant if you have a lot of
files to process, so lloconv supports a "server" mode where a background
process provides a way to reuse a single LibreOfficeKit instance to convert
multiple documents.

This works via a Unix domain socket.  Bear in mind that if another user on
the same machine can write to the domain socket you use then they can probably
read your documents by using lloconv, so it's a good idea to create the socket
in its own directory which only you can read (`chmod 700 DIRECTORY`).

If you specify `-s SOCKETPATH` and a document path and a server isn't running
for SOCKETPATH then one will be started in the background.  You can also start
a server explicitly by using `lloconv -l -s SOCKETPATH` first without
specifying a document.

Currently you can't use `-u` and `-s SOCKETPATH` together, which means when
using a server you can convert files from paths, but not files from arbitrary
URLs.

inject-meta
-----------

This program converts a textual document to `.odt` format, adjusts the metadata
fields in the `.odt` file as specified by the command line options, then
converts back to the requested output format.  This allows you to add or modify
metadata fields in a document automatically, but the double conversion may
result in additional unwanted changes to the document.

It's more of a worked example than a usable tool, and isn't built by default.
Run configure with option `--enable-extra-programs` to enable it.

If you're packaging lloconv you probably don't want to install `inject-meta` via
the package (at least not in `/usr/bin` or equivalent).

Building
--------

To build a release you need a C++ compiler, make, and the LOK headers,
then the standard autotools build commands should work:

./configure
make
sudo make install

If you want to build from a git checkout, you'll also need autoconf and
automake installed, and to first run:

autoreconf -fiv

If you want to build against LOK headers outside /usr/include, you can
specify LO_INCLUDE_PATH like so:

./configure LO_INCLUDE_PATH="$HOME/git/libreoffice/master/include"

This allows building on a system where `$HOME/git/libreoffice/master` is a
git checkout and `/usr/include` has no LOK headers.

Bugs
----

Known Bugs
~~~~~~~~~~

LibreOffice seems to be unable to handle a file whose filename isn't valid
in the current locale's encoding (for example, a filename which isn't
valid UTF-8 on a typical modern Linux machine).

Other Issues
~~~~~~~~~~~~

Failure to convert a particular file is much more likely an issue with
LibreOffice or the file itself rather than a lloconv bug.  If you find a file
which lloconv can't convert, before reporting a bug against lloconv please
first check if you can convert it by opening it using the LibreOffice UI and
selecting File->Export.  If this also fails, it's either a bad file or a
LibreOffice bug.

If you find a file which LibreOffice can convert but lloconv can't, please
report that at https://gitlab.com/ojwb/lloconv, ideally with an example file
which reproduces the problem.

Related Software
----------------

If you're looked for a Python API to LibreOfficeKit, take a look at pylokit:
https://github.com/xrmx/pylokit


NEWS
lloconv 6.1.1 (2022-09-27)

* Document more LibreOffice options I've discovered in --help.

* README: Note that LibreOffice requires filenames to have a valid encoding in
  the current locale.

* README: Improve markup.

* configure: Fix handling of `./configure LO_INCLUDE_PATH=...` which was
  previously ignored due to a typo.

* Drop note from `--help` output about LibreOffice options needing 4.3.0rc1
  since we dropped support for LibreOffice < 4.3.0 a couple of years ago.

* The check for upstream builds of LibreOffice is now dynamic and no longer
  hard-codes a list of versions to check for (which required updating with each
  new release).

* Automatically find packaged LibreOffice on 64bit Fedora

* Automatically find LibreOffice on macOS.  Reportedly initialising
  LibreOfficeKit hangs but at least we'll be ready once that is fixed
  by the LibreOffice developers. See
  https://gitlab.com/ojwb/lloconv/-/issues/11.

lloconv 6.1.0 (2018-08-20)

* First actual numbered release.


lloconv.cc
/* lloconv.cc - Convert a document using LibreOfficeKit
 *
 * Copyright (C) 2014,2015,2016,2018 Olly Betts
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 */

#include <config.h>

#include <cstdlib>
#include <cstring>
#include <iostream>

#include <sys/socket.h>
#include <sys/un.h>
#include <errno.h>
#include <poll.h>
#include <fcntl.h>
#include <stdio.h>
#include <string.h>
#include <sysexits.h>
#include <unistd.h>

#include "convert.h"

using namespace std;

static void
usage(ostream& os)
{
    os << "Usage: " << program << " [-u|-s SOCKET_PATH] [-f OUTPUT_FORMAT] [-o OPTIONS] INPUT_FILE OUTPUT_FILE\n";
    os << "       " << program << " -s SOCKET_PATH -l\n\n";
    os << "  -u  INPUT_FILE is a URL\n\n";
    os << "Known values for OUTPUT_FORMAT include:\n";
    os << "  For text documents: doc docx fodt html odt ott pdf txt xhtml\n\n";
    os << "Known OPTIONS include:\n";
    os << "  EmbedImages - embed image data in HTML using <img src=\"data:...\">\n";
    os << "  SkipImages - don't include images\n";
    os << "  SkipHeaderFooter - don't include the header and footer\n";
    os << "  XHTML - use XHTML doctype (different to -f xhtml)\n";
    os << "  xhtmlns=foo - XHTML with HTML tags prefixed by foo:\n";
    os << flush;
}

static const int LISTEN_BACKLOG = 64;

// Automatically start a listener if -s is used there isn't one.
static bool auto_listener = true;

static bool
read_string(int fd, char ** s)
{
    unsigned char ch;
    if (read(fd, &ch, 1) != 1) return false;
    size_t len = ch;
    if (len >= 253) {
	unsigned i = len - 251;
	len = 0;
	while (i-- > 0) {
	    if (read(fd, &ch, 1) != 1) return false;
	    len = (len << 8) | ch;
	}
    }

    if (*s) free(*s);
    *s = (char *)malloc(len + 1);
    if (!*s) return false;

    char * p = *s;
    while (len) {
	ssize_t n = read(fd, p, len);
	if (n <= 0) {
	    // Error or EOF!
	    return false;
	}
	p += n;
	len -= n;
    }
    *p = '\0';

    return true;
}

static ssize_t
write_all(int fd, const char * buf, size_t count)
{
    while (count) {
	ssize_t r = write(fd, buf, count);
	if (r < 0) {
	    if (errno == EINTR) continue;
	    return r;
	}
	buf += r;
	count -= r;
    }
    return 0;
}

bool
write_string(int fd, const string & s)
{
    size_t len = s.size();
    char buf[5];
    size_t buf_len = 0;
    if (len < 253) {
	buf[buf_len++] = static_cast<unsigned char>(len);
    } else if (len < 0x10000) {
	buf[buf_len++] = 253;
	buf[buf_len++] = static_cast<unsigned char>(len >> 8);
	buf[buf_len++] = static_cast<unsigned char>(len);
	abort();
    } else if (len < 0x1000000) {
	buf[buf_len++] = 254;
	buf[buf_len++] = static_cast<unsigned char>(len >> 16);
	buf[buf_len++] = static_cast<unsigned char>(len >> 8);
	buf[buf_len++] = static_cast<unsigned char>(len);
	abort();
    } else {
	buf[buf_len++] = 255;
	buf[buf_len++] = static_cast<unsigned char>(len >> 24);
	buf[buf_len++] = static_cast<unsigned char>(len >> 16);
	buf[buf_len++] = static_cast<unsigned char>(len >> 8);
	buf[buf_len++] = static_cast<unsigned char>(len);
	abort();
    }
    if (buf_len != 1) abort();
    return write_all(fd, buf, buf_len) == 0 && write_all(fd, s.data(), len) == 0;
}

static void
read_params(int fd, char ** format, char ** input,
	    char ** output, char ** options)
{
    read_string(fd, format);
    read_string(fd, input);
    read_string(fd, output);
    read_string(fd, options);
}

static void
write_params(int fd, const char * format, const char * input,
	     const char * output, const char * options)
{
    if (!format) format = "";
    if (!options) options = "";
    write_string(fd, format);
    write_string(fd, input);
    write_string(fd, output);
    write_string(fd, options);
}

static int
read_result(int fd)
{
    char * v = NULL;
    if (!read_string(fd, &v)) return -1;
    int r = atoi(v);
    free(v);
    return r;
}

static void
write_result(int fd, int v)
{
    char buf[32];
    sprintf(buf, "%d", v);
    write_string(fd, buf);
}

static int
llo_daemon(const char * socket_path)
try {
    int sock = socket(AF_UNIX, SOCK_STREAM, 0);
    if (sock < 0) {
	perror("socket");
	return 1;
    }

    struct sockaddr_un my_addr;
    memset(&my_addr, 0, sizeof(struct sockaddr_un));
    my_addr.sun_family = AF_UNIX;
    if (strlen(socket_path) >= sizeof(my_addr.sun_path)) {
	fprintf(stderr, "socket path too long\n");
	return 1;
    }
    strcpy(my_addr.sun_path, socket_path);

    if (bind(sock, (struct sockaddr *)&my_addr, sizeof(my_addr)) < 0) {
	perror("bind");
	return 1;
    }

    void * handle = convert_init();
    if (!handle) {
	return EX_UNAVAILABLE;
    }
    char * format = NULL;
    char * input = NULL;
    char * output = NULL;
    char * options = NULL;

    while (true) {
	if (listen(sock, LISTEN_BACKLOG) < 0) {
	    perror("listen");
	    return 1;
	}

	struct sockaddr_un peer_addr;
	socklen_t peer_addr_size = sizeof(struct sockaddr_un);
	int fd = accept(sock, (struct sockaddr *) &peer_addr, &peer_addr_size);
	if (fd < 0) {
	    perror("accept");
	    return 1;
	}

	read_params(fd, &format, &input, &output, &options);
	if (!format[0]) {
	    free(format);
	    format = NULL;
	}
	if (!options[0]) {
	    free(options);
	    options = NULL;
	}
	// Hard-code that the path is a file not a URL when using a server, at
	// least for now.
	int res = convert(handle, false, input, output, format, options);
	write_result(fd, res);
	close(fd);
    }
    convert_cleanup(handle);
} catch (const exception & e) {
    cerr << program << ": LibreOffice threw exception (" << e.what() << ")\n";
    return 1;
}

static int
llo_daemon_convert(int fd, const char * format, const char * input,
		   const char * output, const char * options)
{
    write_params(fd, format, input, output, options);
    return read_result(fd);
}

int
main(int argc, char **argv)
{
    program = argv[0];

    const char * format = NULL;
    const char * options = NULL;
    bool url = false;
    bool listener = false;
    const char * socket_path = NULL;

    // FIXME: Use getopt() or something.
    ++argv;
    --argc;
    while (argv[0] && argv[0][0] == '-') {
	switch (argv[0][1]) {
	    case '-':
		if (argv[0][2] == '\0') {
		    // End of options.
		    ++argv;
		    --argc;
		    goto last_option;
		}
		if (strcmp(argv[0] + 2, "help") == 0) {
		    usage(cout);
		    exit(0);
		}
		if (strcmp(argv[0] + 2, "version") == 0) {
		    cout << "lloconv - " PACKAGE_STRING "\n";
		    exit(0);
		}
		break;
	    case 'f':
		if (argv[0][2]) {
		    format = argv[0] + 2;
		    ++argv;
		    --argc;
		} else {
		    format = argv[1];
		    argv += 2;
		    argc -= 2;
		}
		continue;
	    case 'o':
		if (argv[0][2]) {
		    options = argv[0] + 2;
		    ++argv;
		    --argc;
		} else {
		    options = argv[1];
		    argv += 2;
		    argc -= 2;
		}
		continue;
	    case 'u':
		if (argv[0][2] != '\0') {
		    break;
		}
		url = true;
		++argv;
		--argc;
		continue;
	    case 'l':
		if (argv[0][2] != '\0') {
		    break;
		}
		listener = true;
		++argv;
		--argc;
		continue;
	    case 's':
		if (argv[0][2]) {
		    socket_path = argv[0] + 2;
		    ++argv;
		    --argc;
		} else {
		    socket_path = argv[1];
		    argv += 2;
		    argc -= 2;
		}
		continue;
	}

	cerr << "Option '" << argv[0] << "' unknown\n\n";
	argc = -1;
	break;
    }
last_option:

    if (listener) {
	if (argc != 0 || format || options || url || !socket_path) {
	    usage(cerr);
	    _Exit(EX_USAGE);
	}

	_Exit(llo_daemon(socket_path));
    }

    if ((url && socket_path) || argc != 2) {
	usage(cerr);
	_Exit(EX_USAGE);
    }

    const char * input = argv[0];
    const char * output = argv[1];

    if (socket_path) {
	int fd = socket(AF_UNIX, SOCK_STREAM, 0);
	if (fd < 0) {
	    perror("socket");
	    _Exit(1);
	}

	struct sockaddr_un my_addr;
	memset(&my_addr, 0, sizeof(struct sockaddr_un));
	my_addr.sun_family = AF_UNIX;
	if (strlen(socket_path) >= sizeof(my_addr.sun_path)) {
	    fprintf(stderr, "socket path too long\n");
	    _Exit(1);
	}
	strcpy(my_addr.sun_path, socket_path);

	if (connect(fd, (struct sockaddr *)&my_addr, sizeof(my_addr)) < 0) {
	    if ((errno != ECONNREFUSED && errno != ENOENT) || !auto_listener) {
		perror("connect");
		_Exit(1);
	    }
	    if (errno == ECONNREFUSED)
		unlink(socket_path);
	    pid_t child = fork();
	    if (child == -1) {
		perror("fork");
		_Exit(1);
	    }
	    if (child == 0) {
		_Exit(llo_daemon(socket_path));
	    }
	    // FIXME: Actually wait for daemon to start.
	    sleep(1);
	    if (connect(fd, (struct sockaddr *)&my_addr, sizeof(my_addr)) < 0) {
		perror("connect");
		_Exit(1);
	    }
	}

	int rc = llo_daemon_convert(fd, format, input, output, options);
	_Exit(rc);
    }

    void * handle = convert_init();
    if (!handle) {
	_Exit(EX_UNAVAILABLE);
    }
    int rc = convert(handle, url, input, output, format, options);
    convert_cleanup(handle);

    // Avoid segfault from LibreOffice by terminating swiftly.
    _Exit(rc);
}


urlencode.cc
/* @file urlencode.cc
 * @brief URL encoding as described by RFC3986.
 */
/* Copyright (C) 2011,2014,2015,2016 Olly Betts
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to
 * deal in the Software without restriction, including without limitation the
 * rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
 * sell copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
 * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
 * IN THE SOFTWARE.
 */

#include <config.h>

#include "urlencode.h"

#include <cctype>
#include <cstring>
#include <string>

using namespace std;

void
url_encode_(string & res, const char * p, size_t len, const char * safe)
{
    while (len--) {
	unsigned char ch = *p++;
	if (isalnum(ch) || strchr(safe, ch)) {
	    // Unreserved by RFC3986.
	    res += ch;
	} else {
	    // RFC3986 says we "should" encode as upper case hex digits.
	    res += '%';
	    res += "0123456789ABCDEF"[ch >> 4];
	    res += "0123456789ABCDEF"[ch & 0x0f];
	}
    }
}


inject-meta.cc
/* inject-meta.cc - Inject meta-data into a document
 *
 * Provides an example of how to perform multiple conversions.
 *
 * Copyright (C) 2014,2015,2018 Olly Betts
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 */

#include <config.h>

#include <cerrno>
#include <cstdio>
#include <cstdlib>
#include <cstring>
#include <iostream>
#include <map>
#include <string>

#include <fcntl.h>
#include <sysexits.h>
#include <unistd.h>

#include "convert.h"

using namespace std;

// Create a temporary directory called ${TMPDIR-/tmp} + this with XXXXXX
// replaced.
#define TEMP_DIR_TEMPLATE "/inject-meta-XXXXXX"

static void
usage(ostream& os)
{
    os << "Usage: " << program << " -mNAME=VALUE[...] INPUT_FILE OUTPUT_FILE\n\n";
    os << flush;
}

static void
write_xml_tag(FILE * out, const char * tag, const char * content)
{
    fprintf(out, "<%s>", tag);
    for (const char * p = content; *p; ++p) {
	switch (*p) {
	    case '&':
		fputs("&amp;", out);
		break;
	    case '<':
		fputs("&lt;", out);
		break;
	    case '>':
		fputs("&gt;", out);
		break;
	    default:
		putc(*p, out);
		break;
	}
    }
    fprintf(out, "</%s>", tag);
}

int
main(int argc, char **argv)
{
    program = argv[0];

    map<string, const char *> meta;

    // FIXME: Use getopt() or something.
    ++argv;
    --argc;
    while (argv[0] && argv[0][0] == '-') {
	switch (argv[0][1]) {
	    case '-':
		if (argv[0][2] == '\0') {
		    // End of options.
		    ++argv;
		    --argc;
		    goto last_option;
		}
		if (strcmp(argv[0] + 2, "help") == 0) {
		    usage(cout);
		    exit(0);
		}
		if (strcmp(argv[0] + 2, "version") == 0) {
		    cout << "inject-meta - " PACKAGE_STRING "\n";
		    exit(0);
		}
		break;
	    case 'm': {
		const char * tag = argv[0] + 2;
		const char * eq = strchr(tag, '=');
		if (!eq) {
		    cerr << "Option '" << argv[0] << "' missing '='\n\n";
		    usage(cerr);
		    _Exit(EX_USAGE);
		}
		meta[string(tag, eq - tag)] = eq + 1;
		++argv;
		--argc;
		continue;
	    }
	}

	cerr << "Option '" << argv[0] << "' unknown\n\n";
	argc = -1;
	break;
    }
last_option:

    if (argc != 2) {
	usage(cerr);
	_Exit(EX_USAGE);
    }

    const char * input = argv[0];
    const char * output = argv[1];

    void * handle = convert_init();

    // Create a temporary directory.
    const char * p = getenv("TMPDIR");
    if (!p) p = "/tmp";
    char * dir_template = new char[sizeof(TEMP_DIR_TEMPLATE) + strlen(p)];
    strcpy(dir_template, p);
    strcat(dir_template, TEMP_DIR_TEMPLATE);
    p = mkdtemp(dir_template);
    if (!p) {
	cerr << program << ": mkdtemp() failed (" << strerror(errno) << ")\n";
	_Exit(1);
    }
    string tmpdir(p);

    string odt = tmpdir + "/tmp.odt";
    int rc = convert(handle, false, input, odt.c_str());

    if (!rc) {
	int cwd_fd = open(".", O_RDONLY);
	if (cwd_fd < 0) {
	    cerr << program << ": open(\".\") failed (" << strerror(errno) << ")\n";
	    _Exit(1);
	}
	if (chdir(tmpdir.c_str()) < 0) {
	    cerr << program << ": chdir() failed (" << strerror(errno) << ")\n";
	    _Exit(1);
	}
	FILE * f = popen("unzip -p tmp.odt meta.xml", "r");
	if (!f) {
	    cerr << program << ": popen() failed (" << strerror(errno) << ")\n";
	    _Exit(1);
	}

	FILE * out = fopen("meta.xml", "w");
	if (!out) {
	    cerr << program << ": fopen() failed (" << strerror(errno) << ")\n";
	    _Exit(1);
	}

	char *line = NULL;
	size_t len = 0;
	ssize_t c;

	bool in_meta = false;
	while ((c = getdelim(&line, &len, '>', f)) != -1) {
	    if (in_meta) {
		if (strncmp(line, "</office:meta>", sizeof("</office:meta>") - 1) == 0) {
		    in_meta = false;
		    map<string, const char *>::const_iterator i;
		    for (i = meta.begin(); i != meta.end(); ++i) {
			const char * tag = i->first.c_str();
			write_xml_tag(out, tag, i->second);
		    }
		    meta.clear();
		} else if (line[0] == '<') {
		    map<string, const char *>::iterator i;
		    for (i = meta.begin(); i != meta.end(); ++i) {
			const char * tag = i->first.c_str();
			if (strncmp(line + 1, tag, i->first.size()) == 0 &&
			    line[1 + i->first.size()] == '>') {
			    write_xml_tag(out, tag, i->second);
			    meta.erase(i);
			    if ((c = getdelim(&line, &len, '>', f)) == -1) break;
			    goto next;
			}
		    }
		}
	    } else {
		if (strncmp(line, "<office:meta>", sizeof("<office:meta>") - 1) == 0) {
		    in_meta = true;
		}
	    }
	    fwrite(line, c, 1, out);
next:;
	}

	if (fclose(out) < 0) {
	    cerr << program << ": fclose() failed (" << strerror(errno) << ")\n";
	    _Exit(1);
	}

	if (pclose(f) < 0) {
	    cerr << program << ": pclose() failed (" << strerror(errno) << ")\n";
	    _Exit(1);
	}

	int r = system("zip tmp.odt meta.xml");
	unlink("meta.xml");
	if (!WIFEXITED(r) || WEXITSTATUS(r) != 0) {
	    cerr << program << ": failed to run zip command\n";
	    _Exit(1);
	}

	if (fchdir(cwd_fd) < 0) {
	    cerr << program << ": fchdir() failed (" << strerror(errno) << ")\n";
	    _Exit(1);
	}
	rc = convert(handle, false, odt.c_str(), output);
	unlink(odt.c_str());
    }
    rmdir(tmpdir.c_str());
    convert_cleanup(handle);

    // Avoid segfault from LibreOffice by terminating swiftly.
    _Exit(rc);
}


convert.cc
/* convert.cc - Convert documents using LibreOfficeKit
 *
 * Copyright (C) 2014-2022 Olly Betts
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 */

#include <config.h>

#include "convert.h"

#include <climits>
#include <cstdlib>
#include <exception>
#include <iostream>

#include <sys/types.h>
#include <sys/stat.h>
#include <dirent.h>
#include <unistd.h>
#include <sysexits.h>

#include <LibreOfficeKit/LibreOfficeKit.hxx>

#include "urlencode.h"

using namespace std;
using namespace lok;

// Install location for Debian packages (also Fedora on 32-bit architectures):
#define LO_PATH_DEBIAN "/usr/lib/libreoffice/program"

// Install location for Fedora packages on 64-bit architectures:
#define LO_PATH_FEDORA64 "/usr/lib64/libreoffice/program"

// Install location on macOS.  May not actually work there currently though,
// see: https://gitlab.com/ojwb/lloconv/-/issues/11
#define LO_PATH_MACOS "/Applications/LibreOffice.app/Contents/Frameworks"

const char * program = "<program>";

// Find a LibreOffice installation to use.
static const char *
get_lo_path()
{
    const char * lo_path = getenv("LO_PATH");
    if (lo_path) return lo_path;

    struct stat sb;
#define CHECK_DIR(P) if (stat(P"/versionrc", &sb) == 0 && S_ISREG(sb.st_mode)) return P
#ifdef __APPLE__
    CHECK_DIR(LO_PATH_MACOS);
#else
    CHECK_DIR(LO_PATH_DEBIAN);
    if (sizeof(void*) > 4) {
	CHECK_DIR(LO_PATH_FEDORA64);
    }
#endif

    // Check install locations for .deb files from libreoffice.org,
    // e.g. /opt/libreoffice6.3/program
    DIR* opt = opendir("/opt");
    if (opt) {
	// We require at least LibreOffice 4.3.
	unsigned long best_major = 4;
	unsigned long best_minor = 2;
	static string best_rc;
	struct dirent* d;
	while ((d = readdir(opt))) {
#ifdef DT_DIR
	    // Opportunistically skip non-directories if we can spot them
	    // just by looking at d_type.
	    if (d->d_type != DT_DIR && d->d_type != DT_UNKNOWN) {
		continue;
	    }
#endif
	    if (memcmp(d->d_name, "libreoffice", strlen("libreoffice")) != 0) {
		continue;
	    }

	    char* p = d->d_name + strlen("libreoffice");
	    unsigned long major = strtoul(p, &p, 10);
	    if (major == ULONG_MAX) continue;
	    unsigned long minor = 0;
	    if (*p == '.') {
		minor = strtoul(p + 1, &p, 10);
		if (minor == ULONG_MAX) continue;

		string rc = "/opt/";
		rc += d->d_name;
		rc += "/program";
		if (stat((rc + "/versionrc").c_str(), &sb) != 0 || !S_ISREG(sb.st_mode)) {
		    continue;
		}

		if (major > best_major ||
		    (major == best_major && minor > best_minor)) {
		    best_major = major;
		    best_minor = minor;
		    best_rc = std::move(rc);
		}
	    }
	}
	closedir(opt);
	if (!best_rc.empty()) {
	    return best_rc.c_str();
	}
    }

    cerr << program << ": LibreOffice install not found\n"
	"Set LO_PATH in the environment to the 'program' directory - e.g.:\n"
	"LO_PATH=/opt/libreoffice/program\n"
	"export LO_PATH\n";
    _Exit(1);
}

void *
convert_init()
{
    Office * llo = NULL;
    try {
	const char * lo_path = get_lo_path();
	llo = lok_cpp_init(lo_path);
	if (!llo) {
	    cerr << program << ": Failed to initialise LibreOfficeKit\n";
	    return NULL;
	}
	return static_cast<void*>(llo);
    } catch (const exception & e) {
	delete llo;
	cerr << program << ": LibreOfficeKit threw exception (" << e.what() << ")\n";
	return NULL;
    }
}

void
convert_cleanup(void * h_void)
{
    Office * llo = static_cast<Office *>(h_void);
    delete llo;
}

int
convert(void * h_void, bool url,
	const char * input, const char * output,
	const char * format, const char * options)
try {
    if (!h_void) return 1;
    Office * llo = static_cast<Office *>(h_void);

    string input_url;
    if (url) {
	input_url = input;
    } else {
	url_encode_path(input_url, input);
    }
    Document * lodoc = llo->documentLoad(input_url.c_str(), options);
    if (!lodoc) {
	const char * errmsg = llo->getError();
	cerr << program << ": LibreOfficeKit failed to load document (" << errmsg << ")\n";
	return 1;
    }

    string output_url;
    url_encode_path(output_url, output);
    if (!lodoc->saveAs(output_url.c_str(), format, options)) {
	const char * errmsg = llo->getError();
	cerr << program << ": LibreOfficeKit failed to export (" << errmsg << ")\n";
	delete lodoc;
	return 1;
    }

    delete lodoc;

    return 0;
} catch (const exception & e) {
    cerr << program << ": LibreOfficeKit threw exception (" << e.what() << ")\n";
    return 1;
}
